### PR TITLE
Fix graphql api with parameters

### LINKED ISSUE #15527: Grpahql editor query error
### To Reproduce

A serialization failure exception is thrown when passing parameters in the Graphql editor

For example, parameters:
```json
{
  "loadAllMenuSchema": false
}
```

### Expected behavior

```
Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'GraphQL.Inputs'. Path: $.variables | LineNumber: 0 | BytePositionInLine: 889.

NotSupportedException: Deserialization of types without a parameterless constructor, a singular parameterized constructor, or a parameterized constructor annotated with 'JsonConstructorAttribute' is not supported. Type 'GraphQL.Inputs'.
```

### Screenshots
image

### PR DESCRIPTION
**Pull request by hyzx86**
Fixed #15527

### CODE REVIEW

**Comment by hyzx86** [comment]
Hi @MikeAlhayek ，could you please review this PR? 
This is a very serious problem for me ，Thanks!

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs [hunk 1]
```diff
@@ -46,7 +47,12 @@ public override void ConfigureServices(IServiceCollection services)
             services.AddScoped<IPermissionProvider, Permissions>();
             services.AddTransient<INavigationProvider, AdminMenu>();
             services.AddSingleton<GraphQLMiddleware>();
-            services.AddGraphQL(builder => builder.AddSystemTextJson());
+
+            services.AddGraphQL(builder => builder.AddSystemTextJson(options =>
+            {
+                // Common types of converters are already configured in the assembly "GraphQL.SystemTextJson".
+                options.Converters.Add(GraphQLNamedQueryRequestJsonConverter.Instance);
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs:54]: Shouldn't this need to have also the options we use deserialize document? If you tried to query AdminMenu items for example, will that give you an error? I think here we would need the ContentSeriazilerJsionOptions in addition to your GraphQLNamedQueryRequestJsonConverter

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+namespace OrchardCore.Apis.GraphQL.Json;
+public class GraphQLNamedQueryRequestJsonConverter : JsonConverter<GraphQLNamedQueryRequest>
+{
+    public static GraphQLNamedQueryRequestJsonConverter Instance = new();
+    /// <summary>
+    /// Name for the operation name parameter.
+    /// See https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters
+    /// </summary>
+    private const string OPERATION_NAME_KEY = "operationName";
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:14]: Why the uppercase and the underscore? It should be `OperationNameKey` and the same goes for all of the 
```suggestion
    private const string OperationNameKey = "operationName";
```

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 3]
```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using GraphQL;
+
+namespace OrchardCore.Apis.GraphQL.Json;
+public class GraphQLNamedQueryRequestJsonConverter : JsonConverter<GraphQLNamedQueryRequest>
+{
+    public static GraphQLNamedQueryRequestJsonConverter Instance = new();
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:9]: readonly. and add a new line at the end.
```suggestion
    public static readonly GraphQLNamedQueryRequestJsonConverter Instance = new();
    
```

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        {
+            writer.WritePropertyName(NAMEDQUERY_KEY);
+            JsonSerializer.Serialize(writer, value.NamedQuery, options);
+        }
+        writer.WriteEndObject();
+    }
+
+    public override GraphQLNamedQueryRequest Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        if (reader.TokenType != JsonTokenType.StartObject)
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:70]: add `{ }`

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            throw new JsonException();
+
+        var request = new GraphQLNamedQueryRequest();
+
+        while (reader.Read())
+        {
+            if (reader.TokenType == JsonTokenType.EndObject)
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:77]: add `{ }` anywhere it is missing

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using GraphQL;
+
+namespace OrchardCore.Apis.GraphQL.Json;
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:6]: new line
```suggestion
namespace OrchardCore.Apis.GraphQL.Json;

```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs (hunk 2 above)
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:14]: > Why the uppercase and the underscore? It should be `OperationNameKey` and the same goes for all of the

These are copied from the `Graphql.SystemJson` package

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs (hunk 1 above)
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs:54]: This is only for dealing with GraphqlRequest objects, and I haven't seen any problems elsewhere so far

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs:60]: updated , I hope I'm not mistaken 🤣

```
            services.AddGraphQL(builder => builder.AddSystemTextJson((options, sp) =>
            {
                // Common types of converters are already configured in the assembly "GraphQL.SystemTextJson".
                options.Converters.Add(GraphQLNamedQueryRequestJsonConverter.Instance);

                var contentSerializerJsonOptions = sp.GetRequiredService<IOptions<ContentSerializerJsonOptions>>().Value;
                foreach (var converter in contentSerializerJsonOptions.SerializerOptions.Converters)
                {
                    options.Converters.Add(converter);
                }
            }));
```

Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs:60]: > This is only for dealing with GraphqlRequest objects, and I haven't seen any problems elsewhere so far

If there is no other issues, then don't need to add `ContentSerializerJsonOptions`. But if you needed to use the ContentSerializerJsonOptions, you may be able to do this instead

```
services.AddGraphQL(builder => builder.AddSystemTextJson((options, sp) =>
{
    // Common types of converters are already configured in the assembly "GraphQL.SystemTextJson".
    options.Converters.Add(GraphQLNamedQueryRequestJsonConverter.Instance);

    options.Merge(contentSerializerJsonOptions.SerializerOptions);
}));
```

Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Startup.cs:60]: 检查了代码，这里应该需要的，因为

https://github.com/OrchardCMS/OrchardCore/blob/f40fa29e44c124304c309f77fb4b6cb1079dfc26/src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs#L171C1-L171C73

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs [hunk 7]
```diff
@@ -171,7 +172,7 @@ private async Task ExecuteAsync(HttpContext context)
             await _serializer.WriteAsync(context.Response.Body, result);
         }
 
-        private static GraphQLNamedQueryRequest CreateRequestFromQueryString(HttpContext context, bool validateQueryKey = false)
+        private GraphQLNamedQueryRequest CreateRequestFromQueryStringAsync(HttpContext context, bool validateQueryKey = false)
```
[the remark this answers is not in the document]
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs]: fixed

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs [hunk 8]
```diff
@@ -89,29 +91,28 @@ private async Task ExecuteAsync(HttpContext context)
 
                     if (mediaType.IsSubsetOf(_jsonMediaType) || mediaType.IsSubsetOf(_graphQlMediaType))
                     {
-
+                        using var sr = new StreamReader(context.Request.Body);
+                        var bodyText = await sr.ReadToEndAsync();
```
[the remark this answers is not in the document]
Comment (hyzx86) [reply src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs]: Done

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+public class GraphQLNamedQueryRequestJsonConverter : JsonConverter<GraphQLNamedQueryRequest>
+{
+    public static readonly GraphQLNamedQueryRequestJsonConverter Instance = new();
+    
+    /// <summary>
+    /// Name for the operation name parameter.
+    /// See https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters
+    /// </summary>
+    private const string OperationNameKey = "operationName";
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:16]: private variable should start with `_` followed by lower-case.

```suggestion
    private const string _operationNameKey = "operationName";
```

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    /// See https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters
+    /// </summary>
+    private const string _variablesKey = "variables";
+
+    /// <summary>
+    /// Name for the extensions parameter.
+    /// See https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters
+    /// </summary>
+    private const string _extensionsKey = "extensions";
+
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:35]: Can you please document this property too?

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    public override void Write(Utf8JsonWriter writer, GraphQLNamedQueryRequest value, JsonSerializerOptions options)
+    {
+        writer.WriteStartObject();
+        if (value.Query != null)
+        {
+            writer.WritePropertyName(_queryKey);
+            writer.WriteStringValue(value.Query);
+        }
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:46]: New line after every }

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+        {
+            writer.WritePropertyName(_namedQueryKey);
+            JsonSerializer.Serialize(writer, value.NamedQuery, options);
+        }
+        writer.WriteEndObject();
+    }
+
+    public override GraphQLNamedQueryRequest Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        if (reader.TokenType != JsonTokenType.StartObject)
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:72]: Please add { } one every condition

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            {
+                return request;
+            }
+
+            if (reader.TokenType != JsonTokenType.PropertyName)
+            {
+                throw new JsonException();
+            }
+
+            string key = reader.GetString()!;
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:89]: var

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+                case _extensionsKey:
+                    request.Extensions = JsonSerializer.Deserialize<Inputs>(ref reader, options);
+                    break;
+                default:
+                    reader.Skip();
+                    break;
+            }
+        }
+
+        //unexpected end of data
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:118]:
```suggestion
        // Unexpected end of data.
```

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    /// Name for the extensions parameter.
+    /// See https://github.com/graphql/graphql-over-http/blob/master/spec/GraphQLOverHTTP.md#request-parameters
+    /// </summary>
+    private const string _extensionsKey = "extensions";
+
+    /// <summary>
+    /// Name for the namedQuery parameter.
+    /// </summary>
+    private const string _namedQueryKey = "namedQuery";
+
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:40]:
```suggestion
```

File: src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+            }
+
+            if (reader.TokenType != JsonTokenType.PropertyName)
+            {
+                throw new JsonException();
+            }
+
+            var key = reader.GetString()!;
+
+            //unexpected end of data
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/Json/GraphQLNamedQueryRequestJsonConverter.cs:101]:
```suggestion
            // Unexpected end of data.
```

**Review by MikeAlhayek — APPROVED** [review]

**Comment by hyzx86** [comment]
Hi @Piedone，  can you please review this PR？Thanks

**Comment by Piedone** [comment]
If Mike is OK with this then I'm OK with it too, or is there anything specific why you think it would be useful for me to review as well?

**Comment by hyzx86** [comment]
> If Mike is OK with this then I'm OK with it too, or is there anything specific why you think it would be useful for me to review as well?

No, I was wondering does a PR need at least two approvals to pass?

**Comment by Piedone** [comment]
Nope, one is enough :).

**Comment by hishamco** [comment]
@MikeAlhayek please mention this PR #12845 in the triage meeting today, I might unable to attend today

**Review by sebastienros — APPROVED** [review]

**Pull request merged**